fix(table-test): make guidance less presciptive (#174)
uber-go/guide: src/test-table.md

# Test Tables

Table-driven tests with [subtests] can be a helpful pattern for writing tests
to avoid duplicating code when the core test logic is repetitive.

If a system under test needs to be tested against _multiple conditions_ where
*only* the **inputs** and **outputs** change, a table-driven test *can* and **should**

> be used to reduce redundancy and improve readability.

Reviewer: couple of nits:
 * I think we can tone it down a little (e.g., there's a lot of emphasis *only*, *can*, *should*)
 * I think it's useful to provide some context for why

}
```

### Avoid Messy Table Tests

Table tests become difficult to read and maintain if there becomes significant
logic forks within the subtest `for` loop. Table tests should **NOT** be used whenever 
there needs to be complex logic added inside this execution loop.

Large complex table tests harm readability and maintainability because test readers will
have difficulty debugging test failures that occur.

Table tests like this should be split into either multiple test tables or multiple 
individual `Test...` functions

> Some general guidelines to follow are:

Reviewer 2: Suggestion:
```suggestion
Some ideals to aim for are:
```

That makes it clear that the "perfect test" will meet the following conditions but there's leeway as the authors deem fit.

* Tests should focus on the narrowest unit of behavior possible

> * Tests should contain no `if` statements/logic if at all possible

Reviewer 2: that might be too strict but i agree with the point.
how about
```suggestion
* Tests should aim to minimize `if` statements and other control flow if at all possible
```

> * Tests should focus on the narrowest unit of behavior possible
> * Tests should contain no `if` statements/logic if at all possible
> * All table fields should be used in all table tests
> * All test logic should be evaluated for all table cases

Reviewer 2: suggestion: make this more instructional and in an active voice:

```suggestion
* Focus on the narrowest unit of behavior
* Minimize `if` statements and other control flow if possible
* Ensure that all table fields are used in all tests
* Ensure that all test logic runs for all table cases
```

> Simple logic, such as `shouldErr` and `expectErr`, that does not significantly impact readability,
> is acceptable as long as the tests are clear and understandable.
> 
> A common clear pattern to avoid is the addition of `if` statements for fields such as mock expectations
> and expected behavior that test too much responsibility for a single test.

Reviewer 2: We can connect this to the previous list of Do's if we put the "avoid" case first:

```suggestion
Concretely, that means that you should avoid
having multiple branching pathways (e.g. `shouldError`, `expectCall`, etc.),
addition of `if` statements for specific mock expectations (e.g. `shouldCallFoo`),
or placing functions inside the table (e.g. `setupMocks func(*FooMock)`).

As an exception, if the test body is short and straightforward,
it's acceptable to have a single branching pathway for success versus failure cases
with a table field like `wantErr` to specify error expectations.
```

For example, the following table test is confusing to read:

_Bad:_
```go
func TestComplicatedTable(t *testing.T) {
	tests := []struct {
		give            string
		wantHost        string
		wantPort        string
		wantErr         error
		shouldCallXYZ   bool
		giveXResponse string
		giveXErr      error
		giveYResponse string
		giveYErr      error
	}{
		// ...
	}

	for _, tt := range tests {
		t.Run(tt.give, func(t *testing.T) {
			err := DoThing(tt.stuff)
			if tt.wantErr != nil {

> 				require.EqualError(t, err, tt.wantErr)

Reviewer 2: should return after this

			}
			require.NoError(t, err)
			if tt.shouldCallX {
				xyzMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
			}
			if tt.shouldCallY {

> 				xyzMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)

Reviewer 2: mocks should be set before call

			}
		})
	}
}
```

> This makes the test more difficult to change, to understanding, and to review correctness.

Reviewer 2: typo:

```suggestion
This makes the test more difficult to change, understand, and review correctness of.
```

There are no strict guidelines, but readability and maintainability should always be

> top-of-mind when deciding on use of Table Tests vs. separate tests for multiple inputs/outputs 

Reviewer 2: I'm not sure whether it's `vs.` or `vs`, so I'll just suggest `versus` here:

```suggestion
top-of-mind when deciding on use of Table Tests versus separate tests for multiple inputs/outputs 
```

}
```

> ### Avoid Messy Table Tests

Author: I agree that "Complexity" would probably more clear/concrete from the perspective of a reader trying to understand when to use this.

> Table tests become difficult to read and maintain if there becomes significant
> logic forks within the subtest `for` loop. Table tests should **NOT** be used whenever 
> there needs to be complex logic added inside this execution loop.

Reviewer 3: nit:

- s/if there becomes significant/if there are significant/
- s/within the subtest `for`/within the subtest's `for`/

Alternatively:

```suggestion
Table tests can be difficult to read or maintain if their subtests contain
conditional assertions or other branching logic. Table tests should **NOT** be
used whenever there needs to be complex or conditional logic inside subtests.
```

> Large complex table tests harm readability and maintainability because test readers will

Reviewer 3: ```suggestion
Large, complex table tests harm readability and maintainability because test readers may
```

have difficulty debugging test failures that occur.

Table tests like this should be split into either multiple test tables or multiple 

> individual `Test...` functions

Reviewer 3: ```suggestion
individual `Test...` functions.
```

Some ideals to aim for are:

* Focus on the narrowest unit of behavior
* Minimize `if` statements and other control flow (if possible)
* Ensure that all table fields are used in all tests
* Ensure that all test logic runs for all table cases

Concretely, that means that you should avoid
having multiple branching pathways (e.g. `shouldError`, `expectCall`, etc.),

> addition of `if` statements for specific mock expectations (e.g. `shouldCallFoo`),

Reviewer 3: ```suggestion
using `if` statements for specific mock expectations (e.g. `shouldCallFoo`),
```

As an exception, if the test body is short and straightforward,
it's acceptable to have a single branching pathway for success versus failure cases
with a table field like `wantErr` to specify error expectations.

For example, the following table test is confusing to read:

_Bad:_
```go
func TestComplicatedTable(t *testing.T) {
	tests := []struct {
		give            string
		wantHost        string
		wantPort        string
		wantErr         error
		shouldCallXYZ   bool
		giveXResponse string
		giveXErr      error
		giveYResponse string
		giveYErr      error
	}{
		// ...
	}

	for _, tt := range tests {
		t.Run(tt.give, func(t *testing.T) {
			// setup mocks
			ctrl := gomock.NewController(t)
			xMock := xmock.NewMockX(ctrl)
			yMock := xmock.NewMockY(ctrl)

			// run the test
			err := DoComplexThing(tt.give, xMock, yMock)

			// verify results
			if tt.wantErr != nil {
				require.EqualError(t, err, tt.wantErr)
				return
			}
			require.NoError(t, err)
			if tt.shouldCallX {
				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
			}
			if tt.shouldCallY {
				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
			}
		})
	}
}
```

> This makes the test more difficult to change, understand, and review correctness of.

Reviewer 3: ```suggestion
This complexity makes it more difficult to change, understand, and prove the correctness of the test.
```

> There are no strict guidelines, but readability and maintainability should always be
> top-of-mind when deciding on use of Table Tests versus separate tests for multiple inputs/outputs 
> to a system.

Reviewer 3: ```suggestion
While there are no strict guidelines, readability and maintainability should always be
top-of-mind when deciding between Table Tests versus separate tests for multiple inputs/outputs to a system.
```

}
```

## Avoid Unnecessary Complexity in Table Tests

Table tests can be difficult to read and maintain if the subtests contain conditional 
assertions or other branching logic. Table tests should **NOT** be used whenever 
there needs to be complex or conditional logic inside subtests (i.e. logic inside the `for` loop).

Large, complex table tests harm readability and maintainability because test readers may
have difficulty debugging test failures that occur.

Table tests like this should be split into either multiple test tables or multiple 
individual `Test...` functions.

Some ideals to aim for are:

* Focus on the narrowest unit of behavior

> * Minimize `if` statements and other control flow (if possible)

Reviewer 3: Based on [this comment](https://github.com/uber-go/guide/pull/174#issuecomment-1521829041):

```suggestion
* Minimize "test depth", and avoid conditional assertions (see below)
```

and then expand later, after the list:

> In this context, "test depth" means "within a given test, the number of
> successive assertions that require previous assertions to hold" (similar
> to cyclomatic complexity, except that it also describes prerequisite
> relationships between assertions).
>
> Having "shallower" tests means that there are fewer relationships between
> assertions and, more importantly, that those assertions are less likely
> to be conditional by default.

That might better clarify the actual goal (versus banning conditions in general). I think that this would also allow the guidance to be (1) applied more consistently and (2) more flexible around what a test should look like.

Thoughts?

* Ensure that all table fields are used in all tests
* Ensure that all test logic runs for all table cases

Concretely, that means that you should avoid

> having multiple branching pathways (e.g. `shouldError`, `expectCall`, etc.),

Reviewer: nit: typically it's called `wantErr` (and the example below uses that too)

Reviewer 3: I think the intent was different than `wantErr` (conventionally speaking) - `wantErr` is generally an `error` that you assert (e.g. with `errors.Is`), but `shouldError` is probably a boolean that implies branching for conditional assertion.

Having thought about it a bit more, I think we can probably remove this
paragraph and instead call out the things that are bad in the example below; that would help tie them to a concrete example ("how these things *can be* bad") without necessarily over-prescribing ("these things are *always* bad").

As an exception, if the test body is short and straightforward,
it's acceptable to have a single branching pathway for success versus failure cases
with a table field like `wantErr` to specify error expectations.

For example, the following table test is confusing to read:

**Bad:**
```go
func TestComplicatedTable(t *testing.T) {
	tests := []struct {
		give          string
		want          string
		wantErr       error
		shouldCallX   bool
		shouldCallY   bool
		giveXResponse string
		giveXErr      error
		giveYResponse string
		giveYErr      error
	}{
		// ...
	}

	for _, tt := range tests {
		t.Run(tt.give, func(t *testing.T) {
			// setup mocks
			ctrl := gomock.NewController(t)
			xMock := xmock.NewMockX(ctrl)
			if tt.shouldCallX {
				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
			}
			yMock := ymock.NewMockY(ctrl)
			if tt.shouldCallY {
				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
			}

			got, err := DoComplexThing(tt.give, xMock, yMock)

			// verify results
			if tt.wantErr != nil {
				require.EqualError(t, err, tt.wantErr)
				return
			}
			require.NoError(t, err)
			assert.Equal(t, want, got)
		})
	}
}

> ```

Reviewer: Can we add the "Good" example of the above that allows testing the different cases too -- my worry is we're not seriously considering what the alternative would look like to test the same cases, and whether it's actually better.

Reviewer 3: +1, we should have both good/bad examples where possible, if only to illustrate contrast.

Author: Added a "good" example and tried to emphasize that this isn't a one-size-fits-all situation. The guidance should be more about avoiding complexity and confusion rather than any specific implementation in particular.

* Minimize `if` statements and other control flow (if possible)
* Ensure that all table fields are used in all tests
* Ensure that all test logic runs for all table cases

Concretely, that means that you should avoid
having multiple branching pathways (e.g. `shouldError`, `expectCall`, etc.),
using `if` statements for specific mock expectations (e.g. `shouldCallFoo`),
or placing functions inside the table (e.g. `setupMocks func(*FooMock)`).

As an exception, if the test body is short and straightforward,
it's acceptable to have a single branching pathway for success versus failure cases
with a table field like `wantErr` to specify error expectations.

For example, the following table test is confusing to read:

_Bad:_
```go
func TestComplicatedTable(t *testing.T) {
	tests := []struct {
		give          string
		want          string
		wantErr       error
		shouldCallX   bool
		shouldCallY   bool
		giveXResponse string
		giveXErr      error
		giveYResponse string
		giveYErr      error
	}{
		// ...
	}

	for _, tt := range tests {
		t.Run(tt.give, func(t *testing.T) {
			// setup mocks
			ctrl := gomock.NewController(t)
			xMock := xmock.NewMockX(ctrl)

> 			yMock := xmock.NewMockY(ctrl)
> 
> 			// run the test
> 			got, err := DoComplexThing(tt.give, xMock, yMock)
> 
> 			// verify results
> 			if tt.wantErr != nil {
> 				require.EqualError(t, err, tt.wantErr)
> 				return
> 			}
> 			require.NoError(t, err)
> 			if tt.shouldCallX {
> 				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
> 			}
> 			if tt.shouldCallY {
> 				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
> 			}

Reviewer 2: These EXPECT calls should also be in the "setup mocks" section.
gomock (which is what this looks like) wants expectations set in advance.
```suggestion
			yMock := xmock.NewMockY(ctrl)
			if tt.shouldCallX {
				xMock.EXPECT().Call().Return(tt.giveXResponse, tt.giveXErr)
			}
			if tt.shouldCallY {
				yMock.EXPECT().Call().Return(tt.giveYResponse, tt.giveYErr)
			}

			got, err := DoComplexThing(tt.give, xMock, yMock)

			// verify results
			if tt.wantErr != nil {
				require.ErrorIs(t, err, tt.wantErr)
				return
			}
			require.NoError(t, err)
```
